client/consensus/beefy/src/worker.rs

```diff
@@ -901,6 +896,11 @@ where
 						return;
 					}
 				},
+				// Make sure to pump gossip engine.
+				_ = gossip_engine => {
+					error!(target: LOG_TARGET, "🥩 Gossip engine has terminated, closing worker.");
+					return;
+				},
 				// Process incoming justifications as these can make some in-flight votes obsolete.
 				justif = self.on_demand_justifications.next().fuse() => {
```

Comment:
ok, nvm - it's a method call (`.next()`), so it'll return new future on every loop iteration :) Sorry for the noise :)

Reply:
Afaiu this `await` "magic" is that `async` fn is split into states with each `await` point being a state transition - so in the course of a single iteration of the `select!` loop here (L905 where comment is), if at the start of the `select` the engine was "idle" the `on_demand_justifications.next()` never yields/returns anything just like you say; it is stuck in the `futures::future::pending::<()>().await`.

But that's actually desired, because while `Idle` we don't want to keep trying this `next()`.

`Idle` -> `AwaitingResponse` transition happens outside the `select!` (happens as part of `self.process_new_state()`).

On the next loop iteration a new `select!` is called this time the engine being in `AwaitingResponse` from the get-go.

The confusing part I guess is that it is very not obvious that the `fn next()` "state-machine" state gets recreated on each `select!` - it does not live past the async scope.

To summarize, if I understand correctly:
a) Initial engine state: `Idle` -> `select!` called -> `async fn next()` underneath is a normal `fn next()` implementing a statemachine starting in initial state `"Idle-before-await"` -> it never moves to another state -> but state machine does end once `select!` scope ends -> outside select, `self.process_new_state()` might change engine state to `AwaitingResponse` taking us to `b)`
b) Initial engine state: `AwaitingResponse` -> `select!` called -> `async fn next()` underneath is a normal `fn next()` implementing a statemachine starting in initial state `"AwaitingResponse-before-receiver-await"` -> `receiver` gets the `waker` that will move this statemachine further to `"AwaitingResponse-receiver-got-response"` state when response comes in -> sets new state `Idle` and actually returns smth breaking the outer `select!` -> on next loop iteration we're back to `a)`.

Reply:
Yeah - sorry, I've been missing that the future gets recreated on every select. Sorry for bothering you :)